sympy/combinatorics/rewritingsystem.py

```diff
@@ -359,93 +359,74 @@ def _add_to_automaton(self, rules):
-            else:
-                # Store subwords of current state in an array.
-                current_state_name_array = []
+            elif current_state_type == 'a':
                 next_sub = current_state_name
-                while len(next_sub):
-                    current_state_name_array.append(next_sub)
-                    next_sub = next_sub.subword(1, len(next_sub))
                 # Check if the transition to any new state in posible.
-                for sub_word in current_state_name_array:
-                    if sub_word in accept_states:
+                while len(next_sub):
+                    if next_sub in accept_states:
                         # Add accept states.
-                        if current_state_type == 'a':
-                            for letter in automaton_alphabet:
-                                next = current_state_name*letter
-                                while True:
-                                    if len(next) <= 1:
-                                        if next in self.reduction_automaton.states:
-                                            self.reduction_automaton.states[state].add_transition(letter, next)
-                                        else:
-                                            self.reduction_automaton.states[state].add_transition(letter, 'start')
-                                        break
-                                    else:
-                                        if next in self.reduction_automaton.states:
-                                            self.reduction_automaton.states[state].add_transition(letter, next)
-                                            break
-                                        next = next.subword(1, len(next))
-                        break
+                        for letter in automaton_alphabet:
+                            _next = current_state_name*letter
+                            while len(_next):
+                                if _next in self.reduction_automaton.states:
+                                    self.reduction_automaton.states[state].add_transition(letter, _next)
+                                    break
+                                elif len(_next) == 1:
+                                    self.reduction_automaton.states[state].add_transition(letter, 'start')
+                                _next = _next.subword(1, len(_next))
+                    next_sub = next_sub.subword(1, len(next_sub))
 
         # Add transitions for new states. All symbols used in the automaton are considered here.
-        # Ignore this if `self.automaton_alphabet` = `automaton_alphabet`.
+        # Ignore this if `reduction_automaton.automaton_alphabet` = `automaton_alphabet`.
         if len(self.reduction_automaton.automaton_alphabet) != len(automaton_alphabet):
-            for state in proper_prefixes:
+            for state in accept_states:
```

Comment:
What about the dead states?

Reply:
Dead state transitions could be avoided anyway, right?

Reply:
Ah, I see, you bring it back to start state in `reduce_using_automaton`. Yes, that's fine.